bool done
int8 part